README.md
:warning: Everything between << >> needs to be replaced (remove << >> after replacing)
# CS110 Project Proposal
#  Monke Coin Millionaire 
## CS 110 Final Project
###  Fall, 2021 
### [Assignment Description](https://docs.google.com/document/d/1H4R6yLL7som1lglyXWZ04RvTp_RvRFCCBn6sqv-82ps/edit#)

 [https://github.com/bucs110b1fall21/final-project-jla-coding<repo>](#) 

 https://docs.google.com/presentation/d/1IdEuChcDVK8BtonAIRtgswvMtfufzISVymEA2mJ2Vbo/edit#slide=id.gc4901a2eb_0_6 

### Team:  JLA Coding 
#### Avery vonRoeschlaub, Lukas Ladekarl, Jason Zheng

***

## Project Description *(Software Lead)*
 A life simulator that has a satirical take on the modern cryptocurrency market as well as the way Gen Z live. There will be two forms of currency, regular USD and a singular cryptocurrency, monke coin. The cryptocurrency will move at a random amount while the USD will stay the same, and the user get to move the cash from one currency to the other. The game will start off in your room with your PC which generates crypto and a little cash. The game will start as a cookie clicker as the player tries to make as much money. The User also has to survive, with there being random events and basic human necessities. We will scale by adding more events and more things on the outside to spend cash on.

***    

## User Interface Design *(Front End Specialist)*
![gamescreen](assets/gamescreen.jpg)
The top screen shows and in game text that needs to be said. The right menu shows
the user's dollars and JAL coins with some prescripted news. The bottom menu shows
the food/water inventor; the user can click either hand/pocket to place food/water. The right menu shows the water, food, and tiredness bar and the clothing inventory where the user can decide to put whatever outfit they have that they want to wear. The main game screen starts off in the user's bedroom with their bed that they can sleep in, their computer where they invest in JAL Coins and the small rectangle in the bottom left represents a bed.

![startend](assets/startend.jpg)
Start screen has a start button and the end screen has a try again button, that
is the only buttons the user can click on.
![finalgame](assets/finalgame.jpg)
Has a status bar with a interactable computer sink and a background. The player is animated.
![finalendscreen](assets/finalendscreen.jpeg)
Decided not to have a start screen at the end and only have a end screen
* For example, if your program has a start screen, game screen, and game over screen, you should include a wireframe / screenshot / drawing of each one and a short description of the components
* << You should also have a screenshot of each screen for your final GUI >>

***        

## Program Design *(Backend Specialist)*
* Non-Standard libraries
  * None used, although one was planned
* Class Interface Design
  * ![class diagram](assets/IMG_3287.jpg)
* Classes
  * Button - creates a button that has dimentions and text along with selected and deselected colors. Can be w\either selected or unselected
  * ComputerScreen - handles the interactions and display for the computer screen in game, along with managing the buying and selling of crypto
  * Controler - the controler of the game. Runs the game on loop and handles events, level loading, initializations, etc
  * Economy - controls the economy, specifically the rate for crypto to cash. Would also handle stock graphing if it was implemented
  * Interactable - creates interactable objects that trigger events when interacted with along with testing if the player is interacting with it
  * Inventory - handles the inventory and interactions in the inventory
  * InventorySlot - handles individual inventory slots
  * EquipableSlot - unimplemented
  * InventoryItem - superclass of Consumable and Equipable
  * Consumable - creates item that can be consumed - implemented
  * Equipable - creates item that can be equiped by the player - unimplemented
  * Item - unused
  * player - controls the character you control in game. Handles anything involving the player and stats
  * prop - creates an intangable sprite that can be placed anywhere for aesthetics
  * wall - defines an area that the player cannot move into. Also checks for collisions
  
## Project Structure *(Software Lead)*

The Project is broken down into the following file structure:
* main.py
* src
    * Icon, button.py, computer.py, controller.py, economy.py, foldercontents.txt, interactable.py, inventory.py, item.py, level_data.json, player.py, prop.py, sample\_controller.py, wall.py 
* assets
    * BurgerIcon.png, CharacterBack.png, CharacterFront.png, CharacterLeftSide.png, CharacterLeftSide1.png, CharacterLeftSide2.png, CharacterRightSide.png, CharacterRightSide1.png, CharacterRightSide2.png, Chickenbag.png, EMPTY ROOM.png, IMG\_3287.jpg, RedbullCan.png, WacDonaldsBag.png, WaterIcon.png, WonsterEnergy.png, black\_pixel.png, buy.png, cashtocoin.png, characterbackwalking1.png, characterbackwalking1updated.png, characterbackwalking2.png, characterfrontwalking1.png, characterfrontwalking2.png, chart.png, class\_diagram.jpg, coin.png, cointocash.png, computer.png, dollar.png, foldercontents.txt, gamescreen.jpg, main\_character\_back.png, main\_character\_front.png, main\_character\_side\_moving.png, room1 (2021\_11\_27 22\_00\_04 UTC).png, sell.png, sink.png, startend.jpg
* etc
    * foldercontents.txt, Changelog.md

***

## Tasks and Responsibilities *(Software Lead)*
* You must outline the team member roles and who was responsible for each class/method, both individual and collaborative.

### Software Lead - Lukas Ladekarl

 Worked as integration specialist by keeping up with the progress of both team members. Worked on many files, but was not the main contributor to most, except with economy.py. Designed and created all sprites. Created worked on ideas and worked with the team on how to implement them. Was the only team member with experience with Crypto, so contributed to making the game accurate. 

### Front End Specialist -  Jason Zheng

 Front-end lead conducted significant research on implementing a usable inventory system. The inventory was necessary for allowing the user to keep its hunger stat up with consuming food. Also worked on backend parts of the inventory like the "u" removing only and adding food. Also added the actively changing stats on the top left. Implemented most sprites into the game. Worked mainly on player.py and controller.py.

### Back End Specialist - Avery von Roeschlaub

 The back end specialist worked significantly on the interactables. Created the buy menu and all the controls and movement of values that went with it. Made the animations for the character. Created and managed the datafiles that kept the data for sprites and walls. Worked on most files extensively. 

## Testing *(Software Lead)*
*  We tested by announcing our commits to each other and have the other group members look over and check out the commits. 
    *  On November 25 Avery texts: "Added the level data file and a loading thing for it Also I fixed a movement bug." Both Jason and Lukas pull and check the changes and give positive remarks about the changes. 

* Your ATP

| Step                  | Procedure     | Expected Results  | Actual Results |
| ----------------------|:-------------:| -----------------:| -------------- |
|  1  | press movement keys for character sprite (W, A, S, D)  | Character moves  |correct|
|  2  | press E on interactable box | stats change  |correct but bad hitboxes|
|  3  | incorporate hunger and thirst stats | stats change over time  |correct|
|  4  | stats go to 0 | Game End screen  | correct |
|  5  | incorporate cash | stats in dispayed  | correct |
|  2  | press E on interactable box | stats change  |correct |
|  6  | incorporate crypto | stats change randomly in accordance with economy.py |incorrect|
|  7  | press E on computer | opens buy menu  |correct|
|  1  | press movement keys for character sprite (W, A, S, D)  | Character moves with animation |correct|
|  8  | up and down arrow keys  | move percent up and down  |correct|
|  9  | right and left keys | select sell or buy button  | correct |
|  10 | stats go to 0 | Game End screen  | correct |
|  6  | incorporate crypto | rate changes in buy menu |incorrect|
|  11 | pressing v | opens inventory | correct |   
|  12 | pressing enter | confirms purchase  | incorrect |
|  6  | incorporate crypto | rate changes in buy menu |correct|
|  13 | pressing e | closes buy menu | correct |   
|  14 | incorporate number of coins | stat displayed |correct|
|  15 | pressing enter | confirms purchase  | correct |
|  16 | pressing u | buys food  | incorrect |
|  18 | left click over food | removes food and adds to stats  |correct|
|  16 | pressing u | buys food | correct |
   


Changelog.md
# Changelog v0.0.1 at 11/29/2021 4:00 PM:

- added changelog
- added more comments
- added function headers
- created interactable test level
- added sink and computer interactables
- added interactions
- added button for interactions that only activates on initial press
- fixed interaction attempts (R and D were swapped and U and L didn't work)
- added debug mode and debug sprites (currently shows lines for testing interactions)
- added framerate cap and improved movement code to be more understandable and consistant
- implemented basic thirst and hunger mechanics (currently only thirst can be refilled using a sink)
- added height and width to interactables
- added goto method to props
- added price moment for coin
- probably some other stuff I forgot


src/controller.py
import sys
import pygame
import random
import math
import json
from src import player
from src import economy
from src import wall
from src import prop
from src import interactable
from src import computer
from src import button
from src.inventory import Inventory,InventorySlot,EquipableSlot,InventoryItem,Consumable,Equipable

class Controller:
    def __init__(self, width=800, height=600):
        pygame.init()
        
        #self.debug_mode = False #change this to turn debug mode on and off. Feel free to add or remove stuff from debug mode
        self.myfont = pygame.font.SysFont('Calibri', 22)
        self.width = width
        self.height = height
        self.screen = pygame.display.set_mode((self.width, self.height))
        #self.background.fill((100, 100, 100))
        self.background = pygame.Surface(self.screen.get_size()).convert()
        self.background.fill((100, 100, 100))
        pygame.font.init()  
        pygame.key.set_repeat(1, 25)  #this may not actually be needed       
        self.player = player.Player(self)
        self.economy = economy.Economy()
        self.computer_screen = computer.ComputerScreen(self)
        self.inventory = Inventory(self.player,5,5,1)
        
        level_data_ptr = open("src/level_data.json",'r')
        self.level_data = json.load(level_data_ptr) #saves dictionary of levels and their data
        level_data_ptr.close()
        
        self.props = pygame.sprite.Group()
        self.walls = []
        self.interactables = pygame.sprite.Group() #group of interacts in scene
        self.all_sprites = pygame.sprite.Group(tuple(self.props) + tuple(self.interactables) + (self.player,))

        #self.debug_props = pygame.sprite.Group()       
        #if self.debug_mode: #add debug sprites here
            #self.debug_interact_x = prop.Prop("assets/black_pixel.png", self.player.rect.center[0] - self.player.reach, self.player.rect.center[1], self.player.reach*2, 1) #sprite showing interaction zone in x direction
            #self.debug_interact_y = prop.Prop("assets/black_pixel.png", self.player.rect.center[0], self.player.rect.center[1] - self.player.reach, 1, self.player.reach*2) #sprite showing interaction zone in x direction
            #self.debug_props.add(self.debug_interact_x) #add debug sprites
            #self.debug_props.add(self.debug_interact_y)    
        #self.all_sprites = pygame.sprite.Group((self.player,) + tuple(self.props) + tuple(self.interactables) + tuple(self.debug_props)) #group of all sprites in scene

        self.load_level("game_level") #load test level
        self.state = "GAME" #set game to run
    
    def load_level(self, level_name):
        '''
            Loads the specified level from level_data
            args: self, level_name (string)
            returns: None
        '''
        data = self.level_data[level_name] #data = relevent level data
        for prop_data in data["props"]: #for props in prop list, add the prop
            self.props.add(prop.Prop(prop_data[0], prop_data[1], prop_data[2], prop_data[3], prop_data[4]))
        for wall_data in data["walls"]: #for walls in wall list, add the wall
            self.walls.append(wall.Wall(wall_data[0], wall_data[1], wall_data[2], wall_data[3]))
        for interactable_data in data["interactables"]: #for interactables in interactable list, and the interactable
            self.interactables.add(interactable.Interactable(interactable_data[0], interactable_data[1], interactable_data[2], interactable_data[3], interactable_data[4], interactable_data[5]))
        player_data = data["player_data"]
        self.player.goto(player_data[0], player_data[1])
        self.player.face(player_data[2])
        self.all_sprites = pygame.sprite.Group(tuple(self.props) + tuple(self.interactables) + (self.player,))
        self.player.goto(player_data[0], player_data[1]) #move player to correct spot
        self.player.face(player_data[2]) #turn player in correct direction
        #self.all_sprites = pygame.sprite.Group((self.player,) + tuple(self.props) + tuple(self.interactables) + tuple(self.debug_props)) #set all sprites to the new sprite groups (may not be nescesary?)
        beverage = Consumable('assets/WonsterEnergy.png', 1, 0, 10)
        beverage2 = Consumable('assets/RedbullCan.png', 1, 0, 10)
        self.inventory.addItemInv(beverage)
        self.inventory.addItemInv(beverage2)
    
    def unload_level(self):
        '''
            Unloads the level
            args: self
            returns: None
        '''
        for prop in self.props: #kill all props
            prop.kill()
        while walls: #remove all walls
            walls[0].pop()
        for interactable in interactables: #kill all interactables
            interactable.kill()
    
    
    def mainLoop(self):
        '''
            Main game loop
            args: self
            returns: None
        '''
        while True:
            if(self.state == "GAME"):
                self.gameLoop()
            elif(self.state == "GAMEOVER"):
                self.gameOver()
    
    
    def handle_interactions(self, interactions):
        '''
            Handles interactions for objects and what they do based on their ID
            args: self, list [String]
            returns: None
        '''
        for i in interactions:
            if i == 'computer':
                self.computer_screen.toggle_display()
            elif i == 'sink':
                self.player.drink(30)
            else:
                raise ValueError

    def draw_player_stats(self):
        
        self.hunger = self.myfont.render(f"{int(self.player.hunger)}", False, (250,250,250))
        self.thirst = self.myfont.render(f"{int(self.player.thirst)}" , False, (250,250,250))
        self.money = self.myfont.render(f"{int(self.player.money)/100}" , False, (250,250,250))
        self.coins = self.myfont.render(f"{int(1000*self.player.crypto)/1000}" , False, (250,250,250))
        self.hungerimg = pygame.image.load('assets/BurgerIcon.png').convert_alpha()
        self.thirstimg = pygame.image.load('assets/WaterIcon.png').convert_alpha()
        self.moneyimg = pygame.image.load('assets/dollar.png').convert_alpha()
        self.coinimg = pygame.image.load('assets/coin.png').convert_alpha()
        self.screen.blit(self.hunger,(50,30))
        self.screen.blit(self.thirst,(50,60))
        self.screen.blit(self.money,(50,90))
        self.screen.blit(self.coins,(50,120))
        self.screen.blit(self.hungerimg,(20,30))
        self.screen.blit(self.thirstimg,(20,60))
        self.screen.blit(self.moneyimg,(20,90))
        self.screen.blit(self.coinimg,(20,120)) 
        
    def gameLoop(self):
        '''
            Game loop
            args: self
            returns: None
        '''
    
        loop_time = pygame.time.Clock() #keeps track of time since last frame
        prev_key_state = {
            "e": pygame.key.get_pressed()[pygame.K_e],
            "u": pygame.key.get_pressed()[pygame.K_u],
            "v": pygame.key.get_pressed()[pygame.K_v],
            "left": pygame.key.get_pressed()[pygame.K_LEFT],
            "right": pygame.key.get_pressed()[pygame.K_RIGHT],
            "down": pygame.key.get_pressed()[pygame.K_DOWN],
            "up": pygame.key.get_pressed()[pygame.K_UP],
            "return": pygame.key.get_pressed()[pygame.K_RETURN],
            "space": pygame.key.get_pressed()[pygame.K_SPACE]
            } #last press state of key
        loop_time.tick() #sets the time to 0
        while self.state == "GAME":
            dt = loop_time.tick(60) #records dt, the time the last frame took in milliseconds and caps framerate at 60fps
            if(dt > 20): #if lag occurs, print
                print("lag detected")
            if self.player.hunger <= 0 or self.player.thirst <= 0:
                self.state = "GAMEOVER"
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    sys.exit()
            #if key presses are an event it has trouble updating when a key stops being pressed
            pressed = pygame.key.get_pressed() #allows multiple buttons to be active
            if pressed[pygame.K_w]:
                self.player.move('U', self.walls, dt)
            if pressed[pygame.K_s]:
                self.player.move('D', self.walls, dt)
            if pressed[pygame.K_a]:
                self.player.move('L', self.walls, dt)
            if pressed[pygame.K_d]:
                self.player.move('R', self.walls, dt)
            if pressed[pygame.K_v] and not prev_key_state["v"]:
                self.inventory.toggleInventory()
            if pressed[pygame.K_u] and not prev_key_state["u"]:
                food = Consumable('assets/Chickenbag.png', 1, 10, 0)
                food2 = Consumable('assets/WacDonaldsBag.png', 1, 10, 0)
                chance = 0
                chance += random.randrange(1,3)
                self.player.money -= 2000
                if chance == 1:
                    self.inventory.addItemInv(food)
                elif chance == 2:
                    self.inventory.addItemInv(food2)
            if pressed[pygame.K_UP]:
                if self.computer_screen.display_screen:
                    self.computer_screen.increase_button.select()
                    if not prev_key_state["up"]:
                        self.computer_screen.increase_percentage()
            if pressed[pygame.K_DOWN]:
                if self.computer_screen.display_screen:
                    self.computer_screen.decrease_button.select()
                    if not prev_key_state["down"]:
                        self.computer_screen.decrease_percentage()
            if pressed[pygame.K_LEFT] and not prev_key_state["left"]:
                if self.computer_screen.display_screen:
                    self.computer_screen.sell_button.deselect()
                    self.computer_screen.buy_button.select()
            if pressed[pygame.K_RIGHT] and not prev_key_state["right"]:
                if self.computer_screen.display_screen:
                    self.computer_screen.buy_button.deselect()
                    self.computer_screen.sell_button.select()
            if (pressed[pygame.K_RETURN] and not prev_key_state["return"]) or (pressed[pygame.K_SPACE] and not prev_key_state["space"]):
                if self.computer_screen.buttons[2].selected: #buy button
                    self.computer_screen.buy()
                elif self.computer_screen.buttons[3].selected: #sell button
                    self.computer_screen.sell()
            if event.type == pygame.MOUSEBUTTONDOWN and event.button == 3:
                if self.inventory.display_inventory:
                    mouse_pos = pygame.mouse.get_pos()
                    self.inventory.checkSlot(self.screen, mouse_pos)
            if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
                if self.inventory.display_inventory:
                    self.inventory.moveItem(self.screen)
            if event.type == pygame.MOUSEBUTTONUP and event.button == 1:
                if self.inventory.display_inventory:
                    self.inventory.placeItem(self.screen)
            if pressed[pygame.K_e] and not prev_key_state["e"]: #interacts with objects (will not work if e was pressed last frame)
                if self.computer_screen.display_screen:
                    self.computer_screen.toggle_display()
                else:
                    interactions = []
                    for interact in self.interactables: #create list of interactions this frame
                        interaction = interact.attempt_interact(self.player)
                        if interaction:
                            interactions.append(interaction) #add interaction if interaction is not None
                    self.handle_interactions(interactions) #handle interactions
            if not pressed[pygame.K_DOWN]:
                self.computer_screen.decrease_button.deselect()
            if not pressed[pygame.K_UP]:
                self.computer_screen.increase_button.deselect()
            prev_key_state["e"] = pressed[pygame.K_e]
            prev_key_state["u"] = pressed[pygame.K_u]
            prev_key_state["v"] = pressed[pygame.K_v]
            prev_key_state["left"] = pressed[pygame.K_LEFT]
            prev_key_state["right"] = pressed[pygame.K_RIGHT]
            prev_key_state["up"] = pressed[pygame.K_UP]
            prev_key_state["down"] = pressed[pygame.K_DOWN]
            prev_key_state["return"] = pressed[pygame.K_RETURN]
            prev_key_state["space"] = pressed[pygame.K_SPACE]
            if pressed[pygame.K_i]: #for testing (i stands for info, add whatever needs testing)
                print(self.computer_screen.decrease_button.selected)

            # redraw the entire screen
            #if self.debug_mode:
                #self.debug_interact_x.goto(self.player.rect.center[0] - self.player.reach, self.player.rect.center[1])
                #self.debug_interact_y.goto(self.player.rect.center[0], self.player.rect.center[1] - self.player.reach)
            self.all_sprites.update()

            self.player.update_health(dt) #update player hunger and thirst

            self.economy.changerate(dt)
            self.player.update_health(dt) #update player hunger and thirst
            self.screen.blit(self.background, (0, 0))
            self.screen.blit(pygame.transform.scale(self.background, (self.width, self.height)), (0, 0))

            if self.player.hunger == 0 or self.player.thirst == 0:
                self.state = "GAMEOVER"

            self.all_sprites.draw(self.screen) #draw sprites
            self.draw_player_stats()
            self.computer_screen.draw(self.screen)
            self.inventory.draw(self.screen)
            pygame.display.flip() # update the screen
        
    def gameOver(self):
        '''
            Handles gameover state
            args: self
            returns: None
        '''
        self.background.fill((255, 0, 0))
        self.screen.blit(self.background, (0, 0))
        myfont = pygame.font.SysFont(None, 75)
        message = myfont.render('Game Over', False, (0, 0, 0))
        self.screen.blit(message, (250, 250))
        pygame.display.flip()
        while True:
            for event in pygame.event.get(): #make sure you can quit in gameover
                if event.type == pygame.QUIT:
                    sys.exit()
